### PR TITLE
Refactor OptimizeIndexing

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102611
* #102722
* __->__ #100549

This PR decouples the logic necessary to compute bounds on variables
from the logic that uses this info to perform the strenght analysis on
int64 variables. While doing so, it tries to minimize the number of
attributes of the class in favour of local variables.

This class is now accessible from any `LoopBody` object.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov @soumith @anijain2305 @desertfire

### CODE REVIEW

**Comment by Skylion007** [comment]
I don't think you meant to revert the changes to CUTLASS

File: torch/_inductor/bounds.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        self.unbounded_vars = dominated_nodes(
+            node
+            for node in self.loop_body.get_nodes()
+            if node.target in ["load", "reduction"] or "masked_subblock" in node.target
+        )
+        # To access this variable call `get_bounds()`
+        self._bounds: Optional[Dict[torch.fx.Node, ValueRanges]] = {}
+
+    @cache_on_self
+    def get_bounds(self):
```
Comment (eellison) [inline torch/_inductor/bounds.py:33]: if we're doing loop body modification (we are), the cache will be invalid, right ?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -3990,6 +3990,21 @@ def __init__(self, fn, args, var_ranges):
...
+    @cache_on_self
+    def get_nodes(self):
+        all_graphs = itertools.chain(
+            (self.root_block.graph,),
+            (block.graph for block in self.subblocks.values()),
+        )
+        return [node for graph in all_graphs for node in graph.nodes]
+
+    @cache_on_self
+    def bounds(self):
```
Comment (eellison) [inline torch/_inductor/ir.py:4002]: same worry about this one

File: torch/_inductor/optimize_indexing.py [hunk 3]
```diff
@@ -63,247 +36,85 @@ def range_expressable_in_32_bits(range):
...
+    for dominated in dominated_nodes([node], skip_filter):
+        if dominated.target in ["store", "output"]:
+            continue
+
+        if "set_indirect" in dominated.target:
+            idx = int(dominated.target[len("set_indirect") :])
+            indirect_var = indirect_vars[idx]
+
+            # We check that we can compute all the indices it's involved in with int32
+            # TODO The current implementation does not handle the case were an indexing
```
Comment (eellison) [inline torch/_inductor/optimize_indexing.py:62]: where did the new TODO come up ?

**Review by eellison — COMMENTED** [review]
nice !! Couple questions about the caching

File: torch/_inductor/bounds.py (hunk 1 above)
Comment (lezcano) [reply torch/_inductor/bounds.py:33]: Yeah, this should be called once the node is not going to be modified any more. Now, I don't know of any "more secure" way of doing this, as we cannot detect when has the node been modified without a reasonable perf hit.

File: torch/_inductor/optimize_indexing.py (hunk 3 above)
Comment (lezcano) [reply torch/_inductor/optimize_indexing.py:62]: It just occurred to me when having a look at this code. This situation comes up quite often when you have different tensors indexing different dimensions. For example, in the triton code generated in the OP of https://github.com/pytorch/pytorch/issues/100530

File: torch/_inductor/bounds.py (hunk 1 above)
Comment (lezcano) [reply torch/_inductor/bounds.py:33]: Do you reckon we should just recompute it in every call?

**Review by jansel — COMMENTED** [review]
Overall this seems reasonable to me, I will defer to @eellison on the final review as it seems to related to his bounds analysis work.

File: torch/_inductor/bounds.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/bounds.py:102]: Can you add an option to `get_bounds` that will force this to be recomputed ? and add a docstring that its up to the caller to determine whether the cache is valid

**Review by eellison — APPROVED** [review]
LGTM

File: torch/_inductor/bounds.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/bounds.py:102]: Landed bc ur on pto but would be good to address in follow up

**Pull request closed without merging**